File: frontend/src/lib/types/analytics.ts
/**
 * Analytics Types
 */

export interface ApplicationStatistics {
  id: string;
  user_id: string;
  total_submitted: number;
  total_pending: number;
  total_rejected: number;
  total_interviews: number;
  total_offers: number;
  response_rate?: number;
  average_response_time_days?: number;
  success_rate?: number;
  last_updated: string;
}

export interface ApplicationTrend {
  id: string;
  user_id: string;
  date: string;
  applications_submitted: number;
  applications_reviewed: number;
  interviews_scheduled: number;
  rejections_received: number;
  offers_received: number;
}

export interface RoleAnalytics {
  id: string;
  job_title: string;
  application_count: number;
  interview_count: number;
  offer_count: number;
  rejection_count: number;
  last_applied?: string;
  success_rate?: number;
}

export interface CompanyAnalytics {
  id: string;
  company_name: string;
  application_count: number;
  interview_count: number;
  offer_count: number;
  rejection_count: number;
  last_applied?: string;
  success_rate?: number;
}

export interface SourceAnalytics {
  id: string;
  source_name: string;
  application_count: number;
  interview_count: number;
  offer_count: number;
  rejection_count: number;
  success_rate?: number;
}

// Response types

export interface DashboardMetrics {
  total_applications: number;
  pending_applications: number;
  rejected_applications: number;
  interview_scheduled: number;
  offers_received: number;
  response_rate?: number;
  success_rate?: number;
  average_response_time?: number;
}

export interface DashboardResponse {
  metrics: DashboardMetrics;
  top_roles: RoleAnalytics[];
  top_companies: CompanyAnalytics[];
  sources: SourceAnalytics[];
  recent_trends?: ApplicationTrend[];
  last_updated: string;
}

export interface TrendsResponse {
  period: "30days" | "90days" | "all_time";
  trends: ApplicationTrend[];
  total_applications: number;
  average_per_day: number;
  peak_day?: string;
}

export interface TopRolesResponse {
  top_roles: RoleAnalytics[];
  total_unique_roles: number;
}

export interface TopCompaniesResponse {
  top_companies: CompanyAnalytics[];
  total_unique_companies: number;
}

export interface SourceBreakdownResponse {
  sources: SourceAnalytics[];
  total_sources: number;
}

export interface StatusBreakdown {
  status: string;
  count: number;
  percentage: number;
}

export interface DetailedBreakdownResponse {
  by_status: StatusBreakdown[];
  by_title: Record<string, number>;
  by_company: Record<string, number>;
  by_source: Record<string, number>;
  by_experience_level: Record<string, number>;
}

export interface ApplicationInsight {
  insight_type: "strength" | "opportunity" | "warning";
  title: string;
  description: string;
  metric?: number;
}

export interface InsightsResponse {
  insights: ApplicationInsight[];
  generated_at: string;
}

export interface ExportRecord {
  job_title: string;
  company_name: string;
  job_source?: string;
  submission_date: string;
  current_status: string;
  resume_version?: string;
  cover_letter_version?: string;
}

export interface ExportResponse {
  total_records: number;
  records: ExportRecord[];
  exported_at: string;
}

File: backend/app/schemas/analytics.py
"""
Analytics and dashboard schemas
"""

from datetime import datetime, date
from typing import Optional, List, Dict
from uuid import UUID

from pydantic import BaseModel


# Application Statistics Schemas
class ApplicationStatisticsResponse(BaseModel):
    """Application statistics response"""
    id: UUID
    user_id: UUID
    total_submitted: int
    total_pending: int
    total_rejected: int
    total_interviews: int
    total_offers: int
    response_rate: Optional[float] = None
    average_response_time_days: Optional[float] = None
    success_rate: Optional[float] = None
    last_updated: datetime

    class Config:
        from_attributes = True


# Trends Schemas
class ApplicationTrendResponse(BaseModel):
    """Single day trend response"""
    id: UUID
    user_id: UUID
    date: date
    applications_submitted: int
    applications_reviewed: int
    interviews_scheduled: int
    rejections_received: int
    offers_received: int

    class Config:
        from_attributes = True


class TrendsResponse(BaseModel):
    """Trends over time period"""
    period: str  # "30days", "90days", "all_time"
    trends: List[ApplicationTrendResponse]
    total_applications: int
    average_per_day: float
    peak_day: Optional[date] = None


# Role Analytics Schemas
class RoleAnalyticsResponse(BaseModel):
    """Role analytics response"""
    id: UUID
    job_title: str
    application_count: int
    interview_count: int
    offer_count: int
    rejection_count: int
    last_applied: Optional[date] = None
    success_rate: Optional[float] = None

    class Config:
        from_attributes = True


# Company Analytics Schemas
class CompanyAnalyticsResponse(BaseModel):
    """Company analytics response"""
    id: UUID
    company_name: str
    application_count: int
    interview_count: int
    offer_count: int
    rejection_count: int
    last_applied: Optional[date] = None
    success_rate: Optional[float] = None

    class Config:
        from_attributes = True


# Source Analytics Schemas
class SourceAnalyticsResponse(BaseModel):
    """Source analytics response"""
    id: UUID
    source_name: str
    application_count: int
    interview_count: int
    offer_count: int
    rejection_count: int
    success_rate: Optional[float] = None

    class Config:
        from_attributes = True


# Dashboard Response Schemas
class DashboardMetrics(BaseModel):
    """Key metrics for dashboard"""
    total_applications: int
    pending_applications: int
    rejected_applications: int
    interview_scheduled: int
    offers_received: int
    response_rate: Optional[float] = None
    success_rate: Optional[float] = None
    average_response_time: Optional[float] = None


class TopRolesResponse(BaseModel):
    """Top applied-to roles"""
    top_roles: List[RoleAnalyticsResponse]
    total_unique_roles: int


class TopCompaniesResponse(BaseModel):
    """Top companies applied to"""
    top_companies: List[CompanyAnalyticsResponse]
    total_unique_companies: int


class SourceBreakdownResponse(BaseModel):
    """Application breakdown by source"""
    sources: List[SourceAnalyticsResponse]
    total_sources: int


class DashboardResponse(BaseModel):
    """Complete dashboard data"""
    metrics: DashboardMetrics
    top_roles: List[RoleAnalyticsResponse]
    top_companies: List[CompanyAnalyticsResponse]
    sources: List[SourceAnalyticsResponse]
    recent_trends: Optional[List[ApplicationTrendResponse]] = None
    last_updated: datetime


# Export Schemas
class ApplicationExportRecord(BaseModel):
    """Record for CSV export"""
    job_title: str
    company_name: str
    job_source: Optional[str] = None
    submission_date: date
    current_status: str
    resume_version: Optional[str] = None
    cover_letter_version: Optional[str] = None


class ExportResponse(BaseModel):
    """Export response"""
    total_records: int
    records: List[ApplicationExportRecord]
    exported_at: datetime


# Period Request Schema
class AnalyticsPeriodRequest(BaseModel):
    """Request for analytics with time period"""
    period: str = "30days"  # 30days, 90days, all_time


# Status Breakdown Schemas
class StatusBreakdownResponse(BaseModel):
    """Application breakdown by status"""
    status: str
    count: int
    percentage: float


class DetailedBreakdownResponse(BaseModel):
    """Detailed breakdown by status"""
    by_status: List[StatusBreakdownResponse]
    by_title: Dict[str, int]
    by_company: Dict[str, int]
    by_source: Dict[str, int]
    by_experience_level: Dict[str, int]


# Insights Schemas
class ApplicationInsight(BaseModel):
    """Application insight"""
    insight_type: str  # "strength", "opportunity", "warning"
    title: str
    description: str
    metric: Optional[float] = None


class InsightsResponse(BaseModel):
    """Application insights"""
    insights: List[ApplicationInsight]
    generated_at: datetime
